lib/tree.rb

```diff
@@ -16,47 +16,138 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def add(key, value)
-    raise NotImplementedError
+  def add_helper(current, new_node)
+  return new_node if current.nil?
+
+  if new_node.key <= current.key 
+    current.left = add_helper(current.left, new_node)
+  else
+    current.right = add_helper(current.right, new_node)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  return current
+  end
+
+  # Time Complexity: O(log(n))
+  # Space Complexity: O(1)
+  def add(key, value = nil)
+    # raise NotImplementedError
+    new_node = TreeNode.new(key, value)
+    if @root.nil?
+      @root = new_node
+    else
+      add_helper(@root, new_node)
+    end
+  end
+
+  def find_helper(current, key)
+    return nil if current.nil?
+
+    if current.key == key
+      return current.value
+    elsif key < current.key
+      find_helper(current.left, key )
+      
+    else 
+      find_helper(current.right, key )
+    end
+
+  end
+
+  # Time Complexity: O(log(n))
+  # Space Complexity: O(1)
   def find(key)
```

Comment:
👍 The time complexity is O(log n) if the tree is balanced and O(n) otherwise.  The space complexity is the same, due to the recursive call stack.